QEP-338-SIP-sip-incremental-builds.md

```diff
@@ -0,0 +1,64 @@
+# QGIS Enhancement: SIP Incremental Builds
+
+**Date** 2025-03-25
+
+**Author** David Koňařík (@dvdkon)
+
+**Contact** dvdkon at konarici dot cz
+
+**Version** QGIS 3.46 at the latest
+
+# Summary
+
+QGIS uses [SIP](https://github.com/Python-SIP/sip) to generate its Python
+bindings. This means parsing all relevant C++ headers, generating C++ source
+files using SIP, and compiling them. The QGIS build system does this whole
+process every time any relevant header is changed.
+
+Since SIP 5, the parsing and generation steps have [slowed down
+significantly](https://github.com/qgis/QGIS/pull/60291), and compiling the
+resulting C++ has never been fast. This has led some developers to stay with
+SIP 4, which is, however, unmaintained and doesn't work with PyQt6, so this
+will soon stop being an option.
+
+I propose adding the option of building Python bindings incrementally -- per
+each header. This should speed up the Python bindings build during development
+by orders of magnitude.
+
+## Proposed Solution
+
+SIP 6's `sipbuild` would be used as a Python module in a script that generates
+one C++ source file per header file. Currently, `sip-build` generates one file
+per class, optionally concatenating them into some fixed number of files, but
+the SIP Python API is considerably more flexible.
+
+A new CMake variable would be introduced that switches the bindings build
+process from the current `sip-build`-based one to using the new script.
+
+## Deliverables
+
+- Python script that generates Python bindings per-file, rather than
+  per-project.
```

Comment:
Would this python script be released on pypi to potentially attract more users or will it be very QGIS specific?